src/java.base/share/classes/jdk/internal/classfile/impl/verifier/ParserVerifier.java

```diff
@@ -0,0 +1,373 @@
+            }
+        }
+    }
+
+    private void verifyFields(List<VerifyError> errors) {
+        record F(Utf8Entry name, Utf8Entry type) {};
+        var fields = new HashSet<F>();
+        for (var f : classModel.fields()) try {
+            if (!fields.add(new F(f.fieldName(), f.fieldType()))) {
+                errors.add(new VerifyError("Duplicate field name %s with signature %s in %s".formatted(f.fieldName().stringValue(), f.fieldType().stringValue(), toString(classModel))));
+            }
+            verifyFieldName(f.fieldName().stringValue());
+        } catch (VerifyError ve) {
+            errors.add(ve);
+        }
+    }
+
+    private void verifyMethods(List<VerifyError> errors) {
+        record M(Utf8Entry name, Utf8Entry type) {};
+        var methods = new HashSet<M>();
+        for (var m : classModel.methods()) try {
+            if (!methods.add(new M(m.methodName(), m.methodType()))) {
+                errors.add(new VerifyError("Duplicate method name %s with signature %s in %s".formatted(m.methodName().stringValue(), m.methodType().stringValue(), toString(classModel))));
+            }
+            if (m.methodName().equalsString(CLASS_INIT_NAME)
+                    && !m.flags().has(AccessFlag.STATIC)) {
+                errors.add(new VerifyError("Method <clinit> is not static in %s".formatted(toString(classModel))));
+            }
+            if (classModel.flags().has(AccessFlag.INTERFACE)
+                    && m.methodName().equalsString(INIT_NAME)) {
+                errors.add(new VerifyError("Interface cannot have a method named <init> in %s".formatted(toString(classModel))));
+            }
+            verifyMethodName(m.methodName().stringValue());
+        } catch (VerifyError ve) {
+            errors.add(ve);
+        }
+    }
+
+    private void verifyAttributes(ClassFileElement cfe, List<VerifyError> errors) {
+        if (cfe instanceof AttributedElement ae) {
+            var attrNames = new HashSet<String>();
+            for (var a : ae.attributes()) {
+                if (!a.attributeMapper().allowMultiple() && !attrNames.add(a.attributeName())) {
+                    errors.add(new VerifyError("Multiple %s attributes in %s".formatted(a.attributeName(), toString(ae))));
+                }
+                verifyAttribute(ae, a, errors);
+            }
+        }
+        switch (cfe) {
+            case CompoundElement<?> comp -> {
+                for (var e : comp) verifyAttributes(e, errors);
+            }
+            case RecordAttribute ra -> {
+                for(var rc : ra.components()) verifyAttributes(rc, errors);
+            }
+            default -> {}
+        }
+    }
+
+    private void verifyAttribute(AttributedElement ae, Attribute<?> a, List<VerifyError> errors) {
+        int payLoad = ((BoundAttribute)a).payloadLen();
+        if (payLoad != switch (a) {
```

Comment:
Please break this up - e.g. with an intermediate `foundPayload` variable - having a multi-line switch nested as an if condition looks very jarring!

Reply:
Is this method only supposed to check the attribute size? It would be nice perhaps to enhance this to enforce more structural constraints - I added a couple of comments in that direction, but there's many more (e.g. for instance make sure that any entry that morally points to a class/method is of the right kind)

Reply:
Some of the checks don't verify the attributes point to valid cp entries; since CF API is lazy, those entries much be expanded by calling the accessors on Bound attributes.

Reply:
Good point.
Constant pool consistency verification is already covered, however pointing to the right entries is not.
Touching all the attribute values will do the job.
Thanks!